File: src/main/scala/com/kirkwoodwest/scalawoods/GridStatus.scala
package com.kirkwoodwest.scalawoods

import com.bitwig.extension.controller.api.{ClipLauncherSlot, CursorTrack}

//implement GridButton Note
class GridStatus(gridButton: GridButton, val clipLauncherSlot: ClipLauncherSlot, val cursorTrack: CursorTrack) extends GridNote {
  var noteOnStatus: Boolean = false
  clipLauncherSlot.isPlaying.addValueObserver((_) => {
    setStatus()
  })
  clipLauncherSlot.hasContent.addValueObserver((_) => {
    setStatus()
  })
  clipLauncherSlot.isRecording.addValueObserver((_) => {
    setStatus()
  })

//  GridButtonNote(cursorTrack,this)

  override def noteOn(noteOn: Boolean): Unit = {
    this.noteOnStatus = noteOn
    setStatus()
  }

  def setStatus():Unit = {
    if (clipLauncherSlot.isPlaying.get) {
      if(noteOnStatus) gridButton.setState(GridButtonState.ClipPlayingNote)
      else gridButton.setState(GridButtonState.ClipPlaying)
    } else if (clipLauncherSlot.isRecording.get) {
      gridButton.setState(GridButtonState.NoteRecording)
    } else if (clipLauncherSlot.hasContent.get) {
      gridButton.setState(GridButtonState.HasClip)
    } else {
      gridButton.setState(GridButtonState.Off)
    }
  }
}

File: src/main/java/com/kirkwoodwest/launchpadminimk3/GridStatus.java
package com.kirkwoodwest.launchpadminimk3;

import com.bitwig.extension.controller.api.ClipLauncherSlot;
import com.bitwig.extension.controller.api.CursorTrack;
import com.kirkwoodwest.launchpadminimk3.hardware.GridButton;
import com.kirkwoodwest.launchpadminimk3.hardware.GridButtonState;

public class GridStatus implements GridNote {
    private final GridButton gridButton;
    private final ClipLauncherSlot clipLauncherSlot;
    private final CursorTrack cursorTrack;
    private boolean noteOnStatus = false;

    public GridStatus(GridButton gridButton, ClipLauncherSlot clipLauncherSlot, CursorTrack cursorTrack) {
        this.gridButton = gridButton;
        this.clipLauncherSlot = clipLauncherSlot;
        this.cursorTrack = cursorTrack;

        clipLauncherSlot.isPlaying().addValueObserver((b)->setStatus());
        clipLauncherSlot.hasContent().addValueObserver((b)->setStatus());
        clipLauncherSlot.isRecording().addValueObserver((b)->setStatus());

        new GridButtonNote(cursorTrack, this);
    }


    @Override
    public void noteOn(boolean noteOn) {
        this.noteOnStatus = noteOn;
        setStatus();
    }

    private void setStatus() {
        if (clipLauncherSlot.isPlaying().get()) {
            if (noteOnStatus) {
                gridButton.setState(GridButtonState.ClipPlayingNote);
            } else {
                gridButton.setState(GridButtonState.ClipPlaying);
            }
        } else if (clipLauncherSlot.isRecording().get()) {
            gridButton.setState(GridButtonState.NoteRecording);
        } else if (clipLauncherSlot.hasContent().get()) {
            gridButton.setState(GridButtonState.HasClip);
        } else {
            gridButton.setState(GridButtonState.Off);
        }
    }
}
